### DISCUSSION TITLE
How to dispose of AvaloniaUI Application

### DISCUSSION THREAD

**Discussion by matthewcrews** [AvaloniaUI/Avalonia#12052] [Q&A] [answered]
I am using AvaloniuaUI to create a add-in for Excel. I can open a window and close it using the "X" button in the top right of the window. When I go to open it again I get the following error:

```
Can not have multiple active ClassicDesktopStyleApplicationLifetime instances and the previously created one was not disposed
```

What is the correct way to handle this situation? I have searched the API for a way to dispose of the application and have not found one. I have also tried to open the main window again instead of launching the application again but have not found a way to do that either.

Is there any documentation or guides for a scenario where you want the application to stay open and just hide/show the main window?
![image](https://github.com/AvaloniaUI/Avalonia/assets/1593215/ad3fb921-3f97-4947-ad8d-73152bca02ae)

**Comment by kekekeks** [comment, marked as answer]
If you are running avalonia from an addin, you aren't supposed to use the lifetime. Use SetupWithoutStarting and the event loop provided by Excel

**Reply by matthewcrews** [reply]
This is what I have for defining the `MainWindow` and `App`. The `showEditor` function is meant to start the application if it is not running and show the `MainWindow` if it is closed. I cannot figure out how to show the `MainWindow` when it is closed though. With this solution I get an error saying that `Setup` was already run for the `AppBuilder` instance.
![image](https://github.com/AvaloniaUI/Avalonia/assets/1593215/20b81758-5a90-45df-b6f9-b942ab4bf7d3)


```
type MainWindow() as this =
    inherit HostWindow()
    do
        base.Title <- "Counter Example"
        base.Height <- 400.0
        base.Width <- 400.0
        Elmish.Program.mkSimple State.init State.update View.view
        |> Program.withHost this
        |> Program.withConsoleTrace
        |> Program.run

type App() =
    inherit Application()

    override this.Initialize() =
        this.Styles.Add (FluentTheme())
        this.RequestedThemeVariant <- Styling.ThemeVariant.Dark

    override this.OnFrameworkInitializationCompleted() =
        match this.ApplicationLifetime with
        | :? IClassicDesktopStyleApplicationLifetime as desktopLifetime ->
            let mainWindow = MainWindow()
            desktopLifetime.MainWindow <- mainWindow
            desktopLifetime.ShutdownMode <- Controls.ShutdownMode.OnMainWindowClose
        | _ -> ()

let appBuilder =
    AppBuilder
        .Configure<App>()
        .UsePlatformDetect()
        .UseSkia()
        .SetupWithoutStarting()

let showEditor () =
    if Application.Current = null then
        let d (_: Application) (_: string[]) = ()
        let myDelegate = AppBuilder.AppMainDelegate d
        appBuilder.Start(myDelegate, [||])
    else
        // Show MainWindow?
    ()
```

**Reply by kekekeks** [reply]
If you have an external event loop, **you don't control the application lifetime**. You shouldn't be starting the app. Just check if  `Application.Current` is null and initialize Avalonia via `SetupWithoutStarting`. Otherwise just use Avalonia APIs like normal.

**Reply by matthewcrews** [reply]
How then do I show the `MainWindow` if it is not visible? What would I do in this `else` case? Who then actually start the application if I use `SetupWithoutStarting`?

```
let showEditor () =
    if Application.Current = null then
        AppBuilder
            .Configure<App>()
            .UsePlatformDetect()
            .UseSkia()
            .SetupWithoutStarting()
    else
        // Show MainWindow?

    ()
```

**Reply by maxkatz6** [reply]
You don't need to start anything, as there is no application lifetime. As well as there is nothing to dispose, unless you want to scope applications and run multiple with different configurations, but I don't think you need it.

So just create a window and show it `var window = new MainWindow(); window.Show();`.

Avalonia lifetime is a completely optional feature that is useful for developers coming from WPF and expecting deeper integration with the OS (like, currently opened windows, shutdown events and so on).

You, of course, can create your own IApplicationLifetime class and start it from AppBuilder, but Avaloina doesn't require you to do so. It might be useful if you are building a public library with excel integration, I guess.

**Reply by matthewcrews** [reply]
Is this what you're expressing? You patience is appreciated as I try to figure this out 😔.

```
type MainWindow() as this =
    inherit HostWindow()
    do
        base.Title <- "Counter Example"
        base.Height <- 400.0
        base.Width <- 400.0
        Elmish.Program.mkSimple State.init State.update View.view
        |> Program.withHost this
        |> Program.withConsoleTrace
        |> Program.run

let editorWindow = MainWindow()

type App() =
    inherit Application()

    override this.Initialize() =
        this.Styles.Add (FluentTheme())
        this.RequestedThemeVariant <- Styling.ThemeVariant.Dark

    override this.OnFrameworkInitializationCompleted() =
        match this.ApplicationLifetime with
        | :? IClassicDesktopStyleApplicationLifetime as desktopLifetime ->
            desktopLifetime.MainWindow <- editorWindow
            desktopLifetime.ShutdownMode <- Controls.ShutdownMode.OnMainWindowClose
        | _ -> ()

let appBuilder =
    AppBuilder
        .Configure<App>()
        .UsePlatformDetect()
        .UseSkia()
        .SetupWithoutStarting()

let showEditor () =
    editorWindow.Show()
```

When I forego any of the `App` code it does not show anything:

```
type MainWindow() as this =
    inherit HostWindow()
    do
        base.Title <- "Counter Example"
        base.Height <- 400.0
        base.Width <- 400.0
        Elmish.Program.mkSimple State.init State.update View.view
        |> Program.withHost this
        |> Program.withConsoleTrace
        |> Program.run

let showEditor () =
    let editorWindow = MainWindow()
    editorWindow.Show()
```

This returns the following error:

> Unable to locate 'Avalonia.Platform.IWindowingPlatform'

![image](https://github.com/AvaloniaUI/Avalonia/assets/1593215/8e83a8a0-9bdf-45df-8e8c-467eb328ee06)

**Reply by maxkatz6** [reply]
This exception usually means that AppBuilder wasn't setup (i.e. SetupWithoutStarting or Start in normal apps wasn't called).
It's a bit hard for me to follow F# code, but is AppBuilder created once and only once per process?

**Reply by maxkatz6** [reply]
Note, AppBuilder needs to be created once before window is created.

**Reply by matthewcrews** [reply]
It should be built once when the application start.

To clarify, I DO need the `App` type and to have the `AppBuilder` setup code for this to work. Yes?

**Reply by maxkatz6** [reply]
You always need an AppBuilder to setup an Avalonia, yes.

Custom App type is not strictly necessary, as you can create a plain Application instance and add SimpleTheme to its styles. But that's on how it is better for you.
Also, in your code OnFrameworkInitializationCompleted is not needed, as there is no lifetime anymore (ApplicationLifetime is always null).

**Reply by matthewcrews** [reply]
When I use the following code I get this exception:

Exception
> The type initializer for '<StartupCode$6>.$AvaloniaDragAndDrop.App$fsx' threw an exception

Code:
```
type MainWindow() as this =
    inherit HostWindow()
    do
        base.Title <- "Counter Example"
        base.Height <- 400.0
        base.Width <- 400.0
        Elmish.Program.mkSimple State.init State.update View.view
        |> Program.withHost this
        |> Program.withConsoleTrace
        |> Program.run

let editorWindow = MainWindow()

type App() =
    inherit Application()

    override this.Initialize() =
        this.Styles.Add (FluentTheme())
        this.RequestedThemeVariant <- Styling.ThemeVariant.Dark

    override this.OnFrameworkInitializationCompleted() =
        match this.ApplicationLifetime with
        | :? IClassicDesktopStyleApplicationLifetime as desktopLifetime ->
            // let mainWindow = MainWindow()
            // desktopLifetime.MainWindow <- mainWindow
            desktopLifetime.MainWindow <- editorWindow
            desktopLifetime.ShutdownMode <- Controls.ShutdownMode.OnMainWindowClose
        | _ -> ()

let appBuilder =
    AppBuilder
        .Configure<App>()
        .UsePlatformDetect()
        .UseSkia()
        .SetupWithoutStarting()

let showEditor () =
    // let editorWindow = MainWindow()
    editorWindow.Show()
```

I define the `MainWindow` type and then create an instance of it. I then define the `App` type and set the `desktopLifetime` field `MainWindow` equal to the `editorWindow` I instantiated. In the `showEditor` function I am trying to show the `editorWindow`.

**Reply by maxkatz6** [reply]
That's how your initialization should look like:
1. Excel triggers showEditor function
2. If branch to ensure that Application.Current exist. If not - run AppBuilder Setup without any lifetime (as you have a special environment).
3. Create a Window (no need to assign it anywhere)
4. Call Window.Show()

**Reply by matthewcrews** [reply]
The following code works though I would still like to know how to use a "vanilla" application that you mentioned above and not define my own `App` type:

```
type MainWindow() as this =
    inherit HostWindow()
    do
        base.Title <- "Counter Example"
        base.Height <- 400.0
        base.Width <- 400.0
        Elmish.Program.mkSimple State.init State.update View.view
        |> Program.withHost this
        |> Program.withConsoleTrace
        |> Program.run

type App() =
    inherit Application()

    override this.Initialize() =
        this.Styles.Add (FluentTheme())
        this.RequestedThemeVariant <- Styling.ThemeVariant.Dark

let showEditor () =
    if Application.Current = null then
            AppBuilder
                .Configure<App>()
                .UsePlatformDetect()
                .UseSkia()
                .SetupWithoutStarting()
            |> ignore

    let editorWindow = MainWindow()
    editorWindow.Show()
```

**Reply by maxkatz6** [reply]
In c# code it would look like this:
```csharp
       AppBuilder
          .Configure(() => new Application
          {
              Styles = { new SimpleTheme() },
              RequestedThemeVariant = ThemeVariant.Dark
          })
          .UsePlatformDetect()
          .UseSkia()
          .SetupWithoutStarting();
```

**Reply by maxkatz6** [reply]
Minor: UseSkia is not necessary, as it's used by default from the  UsePlatformDetect.

**Reply by matthewcrews** [reply]
Final solution for others:

```
open Elmish
open Avalonia
open Avalonia.Themes.Fluent
open Avalonia.Styling
open Avalonia.FuncUI.Hosts
open Avalonia.FuncUI
open Avalonia.FuncUI.Elmish
// My App types
open AvaloniaDragAndDrop
open AvaloniaDragAndDrop.Domain

type MainWindow() as this =
    inherit HostWindow()
    do
        base.Title <- "Counter Example"
        base.Height <- 400.0
        base.Width <- 400.0
        Program.mkSimple State.init State.update View.view
        |> Program.withHost this
        |> Program.withConsoleTrace
        |> Program.run

let showEditor () =
    if Application.Current = null then
        AppBuilder
            .Configure(fun () ->
                let app = new Application()
                app.Styles.Add (new FluentTheme())
                app.RequestedThemeVariant <- ThemeVariant.Dark
                app    
            )
            .UsePlatformDetect()
            .SetupWithoutStarting()
            |> ignore

    let editorWindow = MainWindow()
    editorWindow.Show()
```

**Reply by kuiperzone** [reply]
Thank you
